roles/jd-client/src/lib/downstream.rs

```diff
@@ -376,12 +376,13 @@ impl DownstreamMiningNode {
         let to_send = to_send.into_values();
         for message in to_send {
             let message = if let Mining::NewExtendedMiningJob(job) = message {
-                let jd = self_mutex.safe_lock(|s| s.jd.clone()).unwrap().unwrap();
-                jd.safe_lock(|jd| jd.coinbase_tx_prefix = job.coinbase_tx_prefix.clone())
-                    .unwrap();
-                jd.safe_lock(|jd| jd.coinbase_tx_suffix = job.coinbase_tx_suffix.clone())
+                if let Some(jd) = self_mutex.safe_lock(|s| s.jd.clone()).unwrap() {
+                    jd.safe_lock(|jd| {
+                        jd.coinbase_tx_prefix = job.coinbase_tx_prefix.clone();
+                        jd.coinbase_tx_suffix = job.coinbase_tx_suffix.clone();
+                    })
```

Comment:
I might have misunderstood reading the PR description, but in this way, is the jd is updating the coinbase tx right after it's initialized as a solo miner? I would guess it's not, until a new job is created by the JDC.

Reply:
if initialized as solo miner, then `self_mutex.safe_lock(|s| s.jd.clone()).unwrap()` is `None` and the coinbase is not updated. If you test dev in order to reproduce the error, you should see that the jdc panics exactly because it is unwrapping a `None`. 
BTW like it is works, but if jdc switches to solo mining the coinbase output is the same, and basically the jdc downstreams are mining for the pool for free. Moreover, for each upstream there should be a different coinbase output, with a special coinbase output reserved for solo mining.
Have I misunderstood something or this is a real issue?

Reply:
# List of coinbase outputs used to build the coinbase tx in case of Solo Mining (as last-resort solution of the pools fallback system)

is literally written in the config. Sorry, didn't read!

Reply:
When JDC does fallback to solo-mining, it should change the coinbase tx output, by using the one specified in config file of the JDC (https://github.com/stratum-mining/stratum/blob/a09084426092a6dfef8862571bc4a0098a05e935/roles/jd-client/config-examples/jdc-config-hosted-example.toml#L33). But if jd is being initialised as None, does it mean it's not going to update the coinbase tx output? Maybe I'm just missing some details

Reply:
> When JDC does fallback to solo-mining, it should change the coinbase tx output, by using the one specified in config file of the JDC
> 
> https://github.com/stratum-mining/stratum/blob/a09084426092a6dfef8862571bc4a0098a05e935/roles/jd-client/config-examples/jdc-config-hosted-example.toml#L33
> But if jd is being initialised as None, does it mean it's not going to update the coinbase tx output? Maybe I'm just missing some details

yes perhaps you are right. I should investigate it!

Reply:
What's the purpose of this segment?
https://github.com/lorbax/stratum/blob/58c70a48e0b1c5fbb9f153700ccf867f425c41b8/roles/jd-client/src/lib/downstream.rs#L379

Reply:
the `jd` field of `DownstreamMiningNode` is an `Option<Arc<Mutex<JobDeclarator>>>` and it is `None` if jd is initialized as solo mining mode ([here](https://github.com/stratum-mining/stratum/blob/dev/roles/jd-client/src/lib/mod.rs#L177) the function `listen_for_downstream_mining` is returning a `DownstreamMiningNode.jd: None`). Before was like this
`
                let jd = self_mutex.safe_lock(|s| s.jd.clone()).unwrap().unwrap();
`
and caused a panic.

Reply:
I meant the purpose of this segment:
```
if let Some(jd) = self_mutex.safe_lock(|s| s.jd.clone()).unwrap() {
                    jd.safe_lock(|jd| {
                        jd.coinbase_tx_prefix = job.coinbase_tx_prefix.clone();
                        jd.coinbase_tx_suffix = job.coinbase_tx_suffix.clone();
                    })
                    .unwrap();
                }
```

Why is it needed to set in that way the jd.coinbase_prefix and suffix? I guess that this is done only in the in which JDC is acting normally with a Pool. But why is that needed?